messaging/src/main/java/org/springframework/security/messaging/context/SecurityContextPropagationChannelInterceptor.java

```diff
@@ -0,0 +1,164 @@
+	 */
+	public SecurityContextPropagationChannelInterceptor() {
+		this(SimpMessageHeaderAccessor.USER_HEADER);
+	}
+
+	/**
+	 * Create a new instance that uses the specified header to populate the
+	 * {@link Authentication}.
+	 * @param authenticationHeaderName the header name to populate the
+	 * {@link Authentication}. Cannot be null.
+	 */
+	public SecurityContextPropagationChannelInterceptor(String authenticationHeaderName) {
+		Assert.notNull(authenticationHeaderName, "authenticationHeaderName cannot be null");
+		this.authenticationHeaderName = authenticationHeaderName;
+	}
+
+	public void setSecurityContextHolderStrategy(SecurityContextHolderStrategy strategy) {
+		this.securityContextHolderStrategy = strategy;
+		this.empty = this.securityContextHolderStrategy.createEmptyContext();
+	}
+
+	@Override
+	public Message<?> preSend(Message<?> message, MessageChannel channel) {
+		Authentication authentication = this.securityContextHolderStrategy.getContext().getAuthentication();
+		if (authentication != null) {
+			return MessageBuilder.fromMessage(message).setHeader(this.authenticationHeaderName, authentication).build();
+		}
+		return message;
+	}
+
+	@Override
+	public Message<?> beforeHandle(Message<?> message, MessageChannel channel, MessageHandler handler) {
+		return postReceive(message, channel);
+	}
+
+	@Override
+	public Message<?> postReceive(Message<?> message, MessageChannel channel) {
+		setup(message);
+		return message;
+	}
+
+	@Override
+	public void afterMessageHandled(Message<?> message, MessageChannel channel, MessageHandler handler, Exception ex) {
+		cleanup(message);
+	}
+
+	private void setup(Message<?> message) {
+		Authentication authentication = message.getHeaders().get(this.authenticationHeaderName, Authentication.class);
+		if (authentication != null) {
+			SecurityContext currentContext = this.securityContextHolderStrategy.getContext();
+			Stack<SecurityContext> contextStack = originalContext.get();
+			if (contextStack == null) {
+				contextStack = new Stack<>();
+				originalContext.set(contextStack);
+			}
+			contextStack.push(currentContext);
+			SecurityContext context = this.securityContextHolderStrategy.createEmptyContext();
+			context.setAuthentication(authentication);
+			this.securityContextHolderStrategy.setContext(context);
+		}
+	}
+
+	/**
+	 * Remove the current {@link Authentication} from the context if an
+	 * {@link #authenticationHeaderName} header value is equal to the current
+	 * {@link Authentication}, and restores one from the stack(if any). Used automatically
+	 * by this interceptor from the
+	 * {@link #afterMessageHandled(Message, MessageChannel, MessageHandler, Exception)}
+	 * where handling operation is finished on the other thread. Must be called explicitly
+	 * downstream after {@link org.springframework.messaging.PollableChannel#receive()}
+	 * operation and processing with the message is done.
+	 * @since 6.2
+	 */
+	public void cleanup(Message<?> message) {
```

Comment:
I'm not really clear on why someone can't call `afterMessageHandled`, for example.

Reply:
The `afterMessageHandled(Message<?> message, MessageChannel channel, MessageHandler handler, Exception ex)` is an interceptor contract and it is called by the framework in specific places.
Yes, one of them is indeed from message handler, when `this.messageHandler.handleMessage(message);` is done.
But in case of `PollableChannel` and that `postReceive()`, respectively, the handler is not involved in the process yet, therefore we cannot `cleanup()` automatically.
The suggestion is to do that manually downstream, where one of the option is like this `handle(mySecurityContextPropagationChannelInterceptor::cleanup)` - just to make it realigned with our Messaging API.

Reply:
I guess I'm unclear since the propagation filter in Spring Messaging doesn't do this same thing. Is there a way that this PR could focus on enabling the existing functionality that Spring Messaging wants to port over? We can consider other use cases in the future.

Reply:
I'm not sure what you are talking about. That was really a main point of this PR: move context propagation support out of Spring Integration and make it available for the whole Spring Messaging.

Reply:
What I'm saying is the Spring Integration interceptor, AFAICT, does not expose a `cleanup` method like this PR does. In the original one, `cleanup` is private.

Reply:
Well, I looked into that one more time and I find it as a but since my current interceptor in Spring Integration doesn't clean up thread local on `postReceive`. That `SecurityContextHolder.clearContext()` will be enough for context, but `EMPTY_CONTEXT` will remain. Therefore I still would like to have this public `cleanup()` for end-user to be called in the end of secured message handling downstream.

I can come up with some extra `Closeable` header to be added to the message, but that doesn't look cleaner or transparent for end-user: still has to be called some way from end-user code.

Reply:
Thanks for the explanation, that's helpful.

> that interceptor ... a plain SecurityContextHolder.clearContext(); is enough

Either one seems problematic, TBH. Again, though, this can be the result of my not understanding how things work on the Integration side of things. My thinking is that an interceptor should be self-sufficient; neither a call to `cleanup` nor a call to `SecurityContextHolder.clearContext()` should be necessary outside the interceptor as that leaks the interceptor's internals.

> The afterMessageHandled() is not called after postReceive(). 

(Since `cleanup` has been mentioned in both threads, I'm bringing this comment :point_up: from the other thread to give us one place to focus.)

I don't think I understand how an interceptor works, then, but I may have missed the full context of this comment. When I use the sample you provided (thank you, again), both `postRecieve` and `afterMessageHandled` are invoked through `PollingConsumer`. Can you give a scenario -- possibly in the sample -- where something other than the interceptor needs to call `cleanup`?

Reply:
Oh! I see now 😄 .
That my `PollingConsumer` indeed has a respective logic:
```
			this.handler.handleMessage(theMessage);
			if (!CollectionUtils.isEmpty(interceptorStack)) {
				triggerAfterMessageHandled(theMessage, null, interceptorStack);
			}
```

I introduced it there 10 years ago. 
My bad.
Yes, removing this clean up from public API.

Sorry for the noise!